### PR TITLE
[WIP] Sonata Exceptions

### PR DESCRIPTION
**Pull request by kunicmarko20**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added some `Class::newMethod` to do great stuff

### Changed

### Deprecated

### Removed

### Fixed

### Security
```

## Subject

Replacing PHP Exception with Sonata ones that implement an interface as a marker.

### CODE REVIEW

**Comment by kunicmarko20** [comment]
There is a lot more to be done here I just need a confirmation that I am going in the right direction, cc @sonata-project/contributors

**Comment by greg0ire** [comment]
Very nice! I would also make these exceptions final.

File: src/Exception/BadConfigValueAlongsideProvider.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @author Marko Kunic <kunicmarko20@gmail.com>
+ */
+class BadConfigValueAlongsideProvider extends InvalidArgumentException implements SonataException
+{
+    /**
+     * @param string $configuration
+     *
+     * @return self
+     */
+    public static function fromConfiguration($configuration)
```
Comment (OskarStark) [inline src/Exception/BadConfigValueAlongsideProvider.php]: return type `: self` ?

Comment (kunicmarko20) [reply src/Exception/BadConfigValueAlongsideProvider.php]: It's on `3.x`, we still support PHP 5.6 😢

File: src/Exception/BadConfigValueAlongsideProvider.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+namespace Sonata\AdminBundle\Exception;
+
+use InvalidArgumentException;
+
+/**
+ * @author Marko Kunic <kunicmarko20@gmail.com>
+ */
+final class BadConfigValueAlongsideProvider extends InvalidArgumentException implements SonataException
```
Comment (sir-kain) [inline src/Exception/BadConfigValueAlongsideProvider.php]: It is more optimal to extend with a `\` like this:
`final class BadConfigValueAlongsideProvider extends \InvalidArgumentException implements SonataException`
To spare php engine from not looking up the class in the current namespace first.

Comment (kunicmarko20) [reply src/Exception/BadConfigValueAlongsideProvider.php]: `\` or `use` statement are the same thing, personally, I do not like seeing `\` in my code.

**Comment by OskarStark** [comment]
You should fix the PR header ;-)

**Comment by kunicmarko20** [comment]
What do you think about a generic `InvalidArgument::fromString` exception? Because if I continue like this we will have a lot of exception classes

**Comment by greg0ire** [comment]
> What do you think about a generic InvalidArgument::fromString exception? Because if I continue like this we will have a lot of exception classes

For exceptions that you're 100% sure are not going to need to be caught, maybe…

**Comment by kunicmarko20** [comment]
I started this all wrong, only the exception that can be caught as said by @greg0ire need a marker. Will redo this at one point, closing now.

**Pull request closed without merging**